Text indicator sometimes is clipped during bounce.

megangardner opened the pull request:
#### 4a9813ddb4606e51b94b3273e1138cd953ef932c
<pre>
Text indicator sometimes is clipped during bounce.
<a href="https://bugs.webkit.org/show_bug.cgi?id=290207">https://bugs.webkit.org/show_bug.cgi?id=290207</a>
<a href="https://rdar.apple.com/147602900">rdar://147602900</a>

Reviewed by Wenson Hsieh.

For some reason the code for what kind of animation that a text
indicator &apos;wanted&apos; was on the TextIndicatorLayer. But if the layer
didn&apos;t exists when asking the question, then it would give a
false negative, meaning that the window wasn&apos;t created large enough
to hold the bounce, meaning that the animation would be clipped.
There&apos;s no reason that this logic shouldn&apos;t just be on the text
indicator itself, so just move that logic and update all the call points.

* Source/WebCore/page/TextIndicator.cpp:
(WebCore::TextIndicator::wantsBounce):
(WebCore::TextIndicator::wantsManualAnimation):
* Source/WebCore/page/TextIndicator.h:
* Source/WebCore/page/cocoa/WebTextIndicatorLayer.h:
* Source/WebCore/page/cocoa/WebTextIndicatorLayer.mm:
(-[WebTextIndicatorLayer _animationDuration]):
(-[WebTextIndicatorLayer present]):
(-[WebTextIndicatorLayer indicatorWantsBounce:]): Deleted.
(-[WebTextIndicatorLayer indicatorWantsManualAnimation:]): Deleted.
* Source/WebCore/page/mac/TextIndicatorWindow.mm:
(WebCore::TextIndicatorWindow::clearTextIndicator):
(WebCore::TextIndicatorWindow::setTextIndicator):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView clearTextIndicator:]):

Canonical link: <a href="https://commits.webkit.org/292612@main">https://commits.webkit.org/292612@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/21efd7c418f348f9021dc7dca0217a523575720d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96532 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16146 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6241 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101604 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47052 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98577 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16442 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24580 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73585 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30815 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99535 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12371 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87301 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53921 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12128 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5085 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46380 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82238 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5179 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103628 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23599 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17211 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82636 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23850 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83341 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82011 "Found 3 new API test failures: /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print-errors, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/close-after-print (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20591 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26656 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4176 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17073 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23562 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28717 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23221 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26701 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24962 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/page/TextIndicator.cpp:
```diff
@@ -131,6 +131,38 @@ RefPtr<TextIndicator> TextIndicator::createWithSelectionInFrame(LocalFrame& fram
     return TextIndicator::create(data);
 }
 
+bool TextIndicator::wantsBounce()
```
```suggestion
bool TextIndicator::wantsBounce() const
```

whsieh commented on Source/WebCore/page/TextIndicator.cpp:
```diff
@@ -131,6 +131,38 @@ RefPtr<TextIndicator> TextIndicator::createWithSelectionInFrame(LocalFrame& fram
+    return false;
+}
+
+bool TextIndicator::wantsManualAnimation()
```
```suggestion
bool TextIndicator::wantsManualAnimation() const
```

whsieh commented on Source/WebCore/page/TextIndicator.h:
```diff
@@ -167,6 +167,9 @@ class TextIndicator : public RefCounted<TextIndicator> {
     void setPresentationTransition(TextIndicatorPresentationTransition transition) { m_data.presentationTransition = transition; }
 
+    bool wantsBounce();
+    bool wantsManualAnimation();
```
```suggestion
    bool wantsBounce() const;
    bool wantsManualAnimation() const;
```

whsieh replied:
I think you need to `WEBCORE_EXPORT` `wantsManualAnimation` as well, since you call it from WK2 on iOS.

whsieh commented on Source/WebCore/page/mac/TextIndicatorWindow.mm:
```diff
@@ -104,8 +104,9 @@ - (BOOL)isFlipped
-    
-    if ([m_textIndicatorLayer indicatorWantsBounce:*m_textIndicator]) {
+
+    RefPtr<TextIndicator> textIndicator = m_textIndicator;
```
`textIndicator` is already passed in as a smart pointer to this method. You can just remove this line.

```suggestion
```

megangardner committed:
Text indicator sometimes is clipped during bounce.
https://bugs.webkit.org/show_bug.cgi?id=290207
rdar://147602900

Reviewed by Wenson Hsieh.

For some reason the code for what kind of animation that a text
indicator 'wanted' was on the TextIndicatorLayer. But if the layer
didn't exists when asking the question, then it would give a
false negative, meaning that the window wasn't created large enough
to hold the bounce, meaning that the animation would be clipped.
There's no reason that this logic shouldn't just be on the text
indicator itself, so just move that logic and update all the call points.

* Source/WebCore/page/TextIndicator.cpp:
(WebCore::TextIndicator::wantsBounce):
(WebCore::TextIndicator::wantsManualAnimation):
* Source/WebCore/page/TextIndicator.h:
* Source/WebCore/page/cocoa/WebTextIndicatorLayer.h:
* Source/WebCore/page/cocoa/WebTextIndicatorLayer.mm:
(-[WebTextIndicatorLayer _animationDuration]):
(-[WebTextIndicatorLayer present]):
(-[WebTextIndicatorLayer indicatorWantsBounce:]): Deleted.
(-[WebTextIndicatorLayer indicatorWantsManualAnimation:]): Deleted.
* Source/WebCore/page/mac/TextIndicatorWindow.mm:
(WebCore::TextIndicatorWindow::clearTextIndicator):
(WebCore::TextIndicatorWindow::setTextIndicator):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView clearTextIndicator:]):

Canonical link: https://commits.webkit.org/292612@main
